storage/remote/queue_manager.go

```diff
@@ -575,11 +576,45 @@ func (t *QueueManager) sendMetadataWithBackoff(ctx context.Context, metadata []p
 	return nil
 }
 
+func isSampleOld(sampleAgeLimit time.Duration, ts int64) bool {
+	if sampleAgeLimit == 0 {
+		// If sampleAgeLimit is unset, then we never skip samples due to their age.
+		return false
+	}
+	limitTs := time.Now().Add(-sampleAgeLimit)
+	sampleTs := timestamp.Time(ts)
+	return sampleTs.Before(limitTs)
+}
+
+func isTimeSeriesOldFilter(sampleAgeLimit time.Duration) func(ts prompb.TimeSeries) bool {
+	return func(ts prompb.TimeSeries) bool {
+		if sampleAgeLimit == 0 {
+			// If sampleAgeLimit is unset, then we never skip samples due to their age.
+			return false
+		}
+		switch {
+		// Only the first element should be set in the series, therefore we only check the first element.
+		case len(ts.Samples) > 0:
+			return isSampleOld(sampleAgeLimit, ts.Samples[0].Timestamp)
+		case len(ts.Histograms) > 0:
+			return isSampleOld(sampleAgeLimit, ts.Histograms[0].Timestamp)
+		case len(ts.Exemplars) > 0:
+			return isSampleOld(sampleAgeLimit, ts.Exemplars[0].Timestamp)
+		default:
+			return false
+		}
+	}
+}
+
 // Append queues a sample to be sent to the remote storage. Blocks until all samples are
 // enqueued on their shards or a shutdown signal is received.
 func (t *QueueManager) Append(samples []record.RefSample) bool {
 outer:
 	for _, s := range samples {
+		if isSampleOld(time.Duration(t.cfg.SampleAgeLimit), s.T) {
+			t.metrics.droppedSamplesTotal.Inc()
```

Comment:
the metrics here are 
```
Total number of samples which were dropped after being read from the WAL 
before being sent via remote write, either via relabelling or unintentionally 
because of an unknown reference ID.
```

We should probably update the help text to include the `too old` reason, or use a new metric. I think also one `level.Debug` line per type and one total per `buildWriteRequest` call for the total # of each data type dropped could be useful. Debug so we don't flood logs by default, I saw you had before but removed a `level.Info` line.

Reply:
We've gone with updating the help text for the original metric.

Furthermore, we've gone with a level.Debug per sample for now. Is the suggestion that we have a counter and log on every call to the QueueManager's Append?

Reply:
Sort of, I think the log line here is still to noisy to keep even on debug. So instead we would not log on every call to Append, I think the correct thing to do is to add a `reason` label to the dropped.

We could keep a debug or warn log for the end of `buildWriteRequest/buildTimeSeries` that logs once with the # of samples/exemplars/etc. dropped due to being too old.

Reply:
> I think the correct thing to do is to add a reason label to the dropped. 

That means making the metric a CounterVec, right?

> We could keep a debug or warn log for the end of buildWriteRequest/buildTimeSeries that logs once with the # of samples/exemplars/etc. dropped due to being too old.

Ok, on it, will check how we can propagate this information upwards here.

Reply:
cc @cstyan For the first part, it's implemented in https://github.com/prometheus/prometheus/pull/13002/commits/c0f1ef96e8945c99cbcff7385c9b9b66f4865593.

The second part means passing either passing the logger a few layers below and keep track of each dropped signal before it's marshalled into a request, or pollute the signature of buildWriteRequest to return these three extra parameters (that are ignored for metadata) just to log this one line.

IMHO, this is an intrusive change for small gain, so I was thinking that maybe it's not worth it? Would like your feedback on it; I've went ahead and implemented this on [<>](https://github.com/prometheus/prometheus/pull/13002/commits/0ecfa0758eadd12f584da7612dc467ac194e9744), but I can revert it if you wish so.